woocommerce/woocommerce-gateway-stripe
# Stripe settings persistence - data layer

## Changes proposed in this Pull Request:

Fixes #1810.

This PR implements the API for saving settings, including:

* The `@wordpress/data` store, incl. hooks.
* The REST controller.
* The `Loadable` components.

I have only added saving support for the "Enable express checkouts" checkbox (excluding the location checkboxes) so there is a setting that can be tested but the PR is still focused mostly on the settings API. Support for saving other settings will be implemented in separate PRs.

![Markup on 2021-09-07 at 16:39:43](https://user-images.githubusercontent.com/1759681/132363882-1bb1900d-e0ec-4a3b-bf44-c5c2e8f32c5e.png)

## Testing instructions

1. Enable UPE preview, UPE, and UPE redesign.
1. Visit the settings page. Notice that as settings are loading:
   * A loading indicator box is displayed over the "Express checkouts" card.
   * The "Save changes" button is disabled.
1. Change the value of the "Enable express checkouts" checkbox.
1. Click "Save changes".
   * Notice the button is disabled and marked as busy while settings are saving.
   * Notice a snackbar notification informing of a successful save is displayed.
1. Refresh the page.
1. Notice that the checkbox value is persisted.
1. Toggle the checkbox, save, and refresh the page once more to confirm both the disabled/enabled status is saved.

## Some resolved struggles with compatibility checks

As part of this PR, due to the compatibility checks verifying compatibility with WC 3.3.0, WP 4.7, and PHP 5.6, I had to tweak some things to get the tests to pass. This includes:

* Deregistering WC Blocks' API endpoint (same reason as in WCPay).
  * I have tweaked this method though to do a `class_exists()` check to support even older versions than WCPay does.
* Removed type declarations from a method header (`bool`).
* Changed a test from `WP_REST_Request::has_param()` to `WP_REST_Request::get_param()` as `has_param` is not available in WP 4.7.
  * The only drawback is that now we can't distinguish between `null` and absent but this was not an issue for the code this PR adds.
